sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -137,15 +195,18 @@ void stop() {
 
   boolean allShardsUpToDate() {
     boolean shardsUpToDate = true;
-    for (ShardRecordsIterator shardRecordsIterator : shardIteratorsMap.values()) {
+    ImmutableMap<String, ShardRecordsIterator> currentShardIterators = shardIteratorsMap.get();
+    for (ShardRecordsIterator shardRecordsIterator : currentShardIterators.values()) {
       shardsUpToDate &= shardRecordsIterator.isUpToDate();
     }
     return shardsUpToDate;
   }
 
   KinesisReaderCheckpoint getCheckpointMark() {
-    return new KinesisReaderCheckpoint(transform(shardIteratorsMap.values(),
+    ImmutableMap<String, ShardRecordsIterator> currentShardIterators = shardIteratorsMap.get();
```

Comment:
If I understand correctly, while this method runs, we may see shardIteratorsMap get modified, potentially an unlimited number of times. Moreover, shardIteratorsMap may have been modified between the last call to advance() and the call to this method. Will the generated checkpoint be correct in that case?

Reply:
Checkpoints are always correct because of following properties:
1) shardIteratorsMap is always in a consistent state, i.e. it is guaranteed to not contain any two shards that are in a parent-child relation. This is a requirement for current logic to properly traverse through splits and merges.
2) All records read from particular shard are removed from the queue in advance() method before shardIterator is removed from shardIteratorsMap. This guarantees that checkpoints will contain particular shardIterator at least as long as there are any unread records from this shard.
So even if shardIteratorsMap is modified an unlimited number of times during any other action, it will always be consistent. The only risk of outdated checkpoint are potential duplicates during restoration.

Reply:
These potential duplicates are what worries me. Checkpointing and resuming from a checkpoint should not lead to data duplication - checkpoint/resume is part of normal operation, rather than just failure recovery. I guess you can partially address this by implementing requiresDeduping() and getRecordId() (not sure if it's already done?) - but ideally, that shouldn't be necessary.

Reply:
I assume we can have potential duplicates in every connector when restoring from outdated checkpoint. I was just thinking about probably usual scenario when you make a checkpoint, go on with reading new records and then at some point something failed and you are restoring from the last checkpoint you have. Then the records read since that checkpoint was created will be duplicated.
If you make a checkpoint, do not read anything and just restore from that checkpoint (does that happen during autoscaling?) then all is fine.

This PR does not change anything regarding Kinesis checkpointing. getCurrentRecordId() is already implemented but requiresDeduping() is not (returns false). We never cared about duplicates but I don't see a problem with changing it to true.